IdentityCore/src/webview/embeddedWebview/ui/mac/MSIDWebviewUIController.m

```diff
@@ -0,0 +1,172 @@
+{
+    self = [super init];
+    if (self)
+    {
+        _context = context;
+    }
+    
+    return self;
+}
+
+- (BOOL)createAndLoadView:(NSError **)error
+{
+    // Get UI container to hold the webview
+    NSWindow *window = [self obtainSignInWindow];
+    NSView *rootView = window.contentView;
+    
+    // Prepare the WKWebView
+    WKWebViewConfiguration *webConfiguration = [WKWebViewConfiguration new];
+    WKWebView *webView = [[WKWebView alloc] initWithFrame:rootView.frame configuration:webConfiguration];
+    [webView setAccessibilityIdentifier:@"MSID_SIGN_IN_WEBVIEW"];
+    
+    // Customize the UI
+    [webView setAutoresizingMask:NSViewHeightSizable | NSViewWidthSizable];
+    _laodingIndicator = [self prepareLoadingIndicator];
+    self.window = window;
+    
+    // Append webview and loading indicator
+    _webView = webView;
+    [rootView addSubview:_webView];
+    [rootView addSubview:_laodingIndicator];
+    
+    return YES;
+}
+
+- (void)presentView
+{
+    [self showWindow:nil];
+}
+
+- (void)dismissWebview:(void (^)(void))completion
+{
+    [self close];
+    completion();
+}
+
+- (void)showLoadingIndicator
+{
+    [_laodingIndicator setHidden:NO];
+    [_laodingIndicator startAnimation:nil];
+    [self.window.contentView setNeedsDisplay:YES];
+}
+
+- (void)dismissLoadingIndicator
+{
+    [_laodingIndicator setHidden:YES];
+    [_laodingIndicator stopAnimation:nil];
+}
+
+- (NSWindow *)obtainSignInWindow
+{
+    NSWindow *mainWindow = [NSApp mainWindow];
+    NSRect windowRect;
+    if (mainWindow)
+    {
+        windowRect = mainWindow.frame;
+    }
+    else
+    {
+        // If we didn't get a main window then center it in the screen
+        windowRect = [[NSScreen mainScreen] frame];
+    }
+    
+    // Calculate the center of the current main window so we can position our window in the center of it
+    NSRect centerRect = [self getCenterRect:windowRect rect2:NSMakeRect(0, 0, DEFAULT_WINDOW_WIDTH, DEFAULT_WINDOW_HEIGHT)];
+    
+    NSWindow *window = [[NSWindow alloc] initWithContentRect:centerRect
+                                                   styleMask:NSTitledWindowMask | NSClosableWindowMask
+                                                     backing:NSBackingStoreBuffered
+                                                       defer:YES];
+    [window setDelegate:self];
+    [window setAccessibilityIdentifier:@"MSID_SIGN_IN_WINDOW"];
+    [window.contentView setAutoresizesSubviews:YES];
+    
+    return window;
+}
+
+- (NSRect)getCenterRect:(NSRect)rect1
+                  rect2:(NSRect)rect2
+{
+    CGFloat x = rect1.origin.x + ((rect1.size.width - rect2.size.width) / 2);
+    CGFloat y = rect1.origin.y + ((rect1.size.height - rect2.size.height) / 2);
+    
+    rect2.origin.x = x;
+    rect2.origin.y = y;
+    
+    return rect2;
+}
+
+// Authentication was cancelled by the user by closing the window
+- (void)windowWillClose:(NSNotification *)notification
+{
```

Comment:
Question:  I'm stealing your latest bits to try them out in cross-plat C++ for macOS.  The popup is coming up and if I close the window manually the windowWillClose method is not being invoked.  Is there some other magic sauce I need to wire up to get this to get called from the popup window so that we can trigger the cancellation flow?  thx :D

Reply:
Seems such an issue exists in current public ADAL as well. I will try to discuss this issue with my team and get back to you later :)

Reply:
Only if you are using pass-in webview.
Seems such an issue exists in current public ADAL as well. I will try to discuss this issue with my team and get back to you later :)

Reply:
Hi @MarkZuber ,
if you pass in your own webview, windowWillClose will not be invoked by design;
if you didn't pass in your own webview and windowWillClose is not invoked, then it is a bug. Do let me know if that is the case. (i have tested it on my side and no issue found)

Reply:
Sure :)